### PR TITLE
fix: sanitize Avro field names on write, respect iceberg-field-name on read

### LINKED ISSUE #2535: avro schema writer does not sanitize field names that violate avro naming rules
## problem

when writing manifest files, iceberg-rust copies iceberg field names verbatim into avro record field names. the avro spec requires names to match `[A-Za-z_][A-Za-z0-9_]*`. partition field names that start with digits (e.g. hash-derived names like `815d3b5701b94c78884835c1bea174bb_day`) produce invalid avro schemas.

java handles this correctly in `TypeToSchema.java`:
```java
String origFieldName = structField.name();
boolean isValidFieldName = AvroSchemaUtil.validAvroName(origFieldName);
String fieldName = isValidFieldName ? origFieldName : AvroSchemaUtil.sanitize(origFieldName);
Schema.Field field = new Schema.Field(fieldName, ...);
if (!isValidFieldName) {
  field.addProp(AvroSchemaUtil.ICEBERG_FIELD_NAME_PROP, origFieldName);
}
```

sanitization rules (from `AvroSchemaUtil.sanitize()`):
- leading digit → prefix with `_` (e.g. `9col` → `_9col`)
- special chars → `_x<hex>` (e.g. `a.b` → `a_x2Eb`)

the original name is preserved in the `iceberg-field-name` avro field property.

## relevant code

`crates/iceberg/src/avro/schema.rs` — `schema_to_avro_schema` uses `field.name.clone()` directly as the avro field name without validation or sanitization.

## impact

- manifests written by iceberg-rust with digit-leading partition field names are invalid avro
- other engines (spark, trino, flink) using strict avro readers will reject these manifests
- any table using hash-based partition transforms can produce such names

## expected behavior

implement the same sanitize-on-write protocol as java:
1. validate name against `[A-Za-z_][A-Za-z0-9_]*`
2. if invalid, sanitize and store original in `iceberg-field-name` property
3. always store `field-id` property (already done)

### PR DESCRIPTION
**Pull request by SreeramGarlapati**
## Summary

Iceberg field names can be anything (`123column`, `field.with.dots`, etc.) but Avro requires names to match `[A-Za-z_][A-Za-z0-9_]*`. Java sanitizes invalid names on write and records the original in the `iceberg-field-name` custom Avro property (`TypeToSchema`, `BuildAvroProjection`). On read, Java recovers the authoritative name by **field-id projection against the table schema**, so in Java that property is write-only provenance rather than something it reads back. iceberg-rust was doing neither half of the write contract — it wrote invalid Avro names directly, and on the standalone Avro→Iceberg schema conversion it had no way to recover the original name.

This causes two interop failures:
1. **Write path**: iceberg-rust produces invalid Avro when field names have leading digits or special chars → strict parsers (Java, Python) reject the file
2. **Read path**: Avro files carrying sanitized names + `iceberg-field-name` are converted back to the *sanitized* name (e.g. `_123column`) instead of the original

### Changes

**Write path** (Iceberg→Avro conversion in `SchemaToAvroSchema::field`):
- Added `is_valid_avro_name()` — checks `[A-Za-z_][A-Za-z0-9_]*`
- Added `sanitize_avro_name()` — follows Java `AvroSchemaUtil.sanitize()`'s escape *scheme*, but is intentionally **ASCII-strict** so output is always a spec-valid Avro name (Java's unicode `Character.isLetter`/`isLetterOrDigit`/`isDigit` can leave names like `café`/`_٠x` that violate Avro's ASCII-only grammar):
  - Leading digit: prefix `_` (e.g., `123col` → `_123col`)
  - Anything outside ASCII alnum/`_`: `_x<HEX>` (e.g., `field.name` → `field_x2Ename`)
  - Operates on UTF-16 code units to match Java's `charAt()` for supplementary chars
- When sanitization is needed: stores original name in `iceberg-field-name` property

**Read path** (Avro→Iceberg conversion in `AvroSchemaToSchema::record`):
- Checks `iceberg-field-name` custom attribute first, falls back to the Avro field name. Note this is deliberately more faithful than Java's `SchemaToType`, which uses `field.name()` directly; iceberg-rust restores the original on the standalone conversion path too, while staying field-id-based for real table reads.

### Java reference

- [`AvroSchemaUtil.sanitize()` / `validAvroName()`](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/avro/AvroSchemaUtil.java#L526-L549)
- [`TypeToSchema.struct()`](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/avro/TypeToSchema.java#L107-L135) — writes `ICEBERG_FIELD_NAME_PROP`
- [`AvroSchemaUtil.ICEBERG_FIELD_NAME_PROP`](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/avro/AvroSchemaUtil.java#L44) — write-only in Java (no read site)

Closes #2535

## Test plan

- [x] `test_is_valid_avro_name` — validates detection of invalid names
- [x] `test_sanitize_avro_name` — ASCII edge cases (match Java for the chars both escape)
- [x] `test_sanitize_avro_name_unicode` — BMP and supplementary char handling (surrogate pairs)
- [x] `test_sanitization_round_trip` — Iceberg→Avro→Iceberg preserves original names
- [x] `test_avro_to_iceberg_uses_iceberg_field_name_property` — reads sanitized schemas back to the original name
- [x] All existing tests pass (no regressions in bidirectional schema conversion tests)

### CODE REVIEW

File: crates/iceberg/src/avro/schema.rs [hunk 1]
```diff
@@ -34,12 +34,83 @@ use crate::{Error, ErrorKind, Result, ensure_data_valid};
...
 const ELEMENT_ID: &str = "element-id";
 const FIELD_ID_PROP: &str = "field-id";
+const ICEBERG_FIELD_NAME_PROP: &str = "iceberg-field-name";
 const KEY_ID: &str = "key-id";
 const VALUE_ID: &str = "value-id";
 const MAP_LOGICAL_TYPE: &str = "map";
 // This const may better to maintain in avro-rs.
 const LOGICAL_TYPE: &str = "logicalType";
 
+fn is_valid_avro_name(name: &str) -> bool {
```
Comment (xanderbailey) [inline crates/iceberg/src/avro/schema.rs:44]: It would be nice if we could use something from https://github.com/apache/avro-rs/blob/4edb1ce1ae1ab5bd3fafb08ca3f622946c01c9fd/avro/src/validator.rs#L4 but it looks like `validate_record_field_name` is `pub(crate)`. Is it work filing an issue upstream to see if we could expose something that would allow us to validate against their implementation?

File: crates/iceberg/src/avro/schema.rs [hunk 2]
```diff
@@ -34,12 +34,83 @@ use crate::{Error, ErrorKind, Result, ensure_data_valid};
...
+
+#[inline]
+fn is_ascii_alpha_u16(c: u16) -> bool {
+    matches!(c, 0x41..=0x5A | 0x61..=0x7A)
+}
+
+#[inline]
+fn is_ascii_digit_u16(c: u16) -> bool {
+    matches!(c, 0x30..=0x39)
+}
```
Comment (xanderbailey) [inline crates/iceberg/src/avro/schema.rs:115]:
```java
    /**
     * Determines if the specified character is a digit.
     * <p>
     * A character is a digit if its general category type, provided
     * by {@code Character.getType(ch)}, is
     * {@code DECIMAL_DIGIT_NUMBER}.
     * <p>
     * Some Unicode character ranges that contain digits:
     * <ul>
     * <li>{@code '\u005Cu0030'} through {@code '\u005Cu0039'},
     *     ISO-LATIN-1 digits ({@code '0'} through {@code '9'})
     * <li>{@code '\u005Cu0660'} through {@code '\u005Cu0669'},
     *     Arabic-Indic digits
     * <li>{@code '\u005Cu06F0'} through {@code '\u005Cu06F9'},
     *     Extended Arabic-Indic digits
     * <li>{@code '\u005Cu0966'} through {@code '\u005Cu096F'},
     *     Devanagari digits
     * <li>{@code '\u005CuFF10'} through {@code '\u005CuFF19'},
     *     Fullwidth digits
     * </ul>
     *
     * Many other character ranges contain digits as well.
     *
     * <p><b>Note:</b> This method cannot handle <a
     * href="#supplementary"> supplementary characters</a>. To support
     * all Unicode characters, including supplementary characters, use
     * the {@link #isDigit(int)} method.
     *
     * @param   ch   the character to be tested.
     * @return  {@code true} if the character is a digit;
     *          {@code false} otherwise.
     * @see     Character#digit(char, int)
     * @see     Character#forDigit(int, int)
     * @see     Character#getType(char)
     */
    public static boolean isDigit(char ch) {
        return isDigit((int)ch);
    }
```
https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/avro/AvroSchemaUtil.java#L551 
Java's isDigit() actually covers more than just ascii digits. This is a pretty niche edge-case. 

Trying to reason about this in my head and I don't think it matters for interop that the representations are identical because both will correctly underscore the field and restore it from the map? Can you check my reasoning here?

**Review by xanderbailey — COMMENTED** [review]
Nice PR, thanks for the contribution, left a couple of questions

File: crates/iceberg/src/avro/schema.rs (hunk 1 above)
Comment (SreeramGarlapati) [reply crates/iceberg/src/avro/schema.rs:44]: good call. i'd keep the local check regardless though — `is_valid_avro_name` is the exact Avro grammar `[A-Za-z_][A-Za-z0-9_]*` in ~6 lines with no coupling to avro-rs internals, and the sanitize half is the iceberg `iceberg-field-name` convention which avro-rs doesn't cover anyway. happy to file an upstream issue to expose `validate_record_field_name` as good-citizen cleanup and link it here — lmk if you'd like that.

File: crates/iceberg/src/avro/schema.rs (hunk 2 above)
Comment (SreeramGarlapati) [reply crates/iceberg/src/avro/schema.rs:115]: your conclusion holds — representational parity with Java isn't needed — but let me correct the reasoning (incl. something I'd have gotten wrong).

this impl is deliberately **not** matching Java's `isDigit`. Java's `validAvroName`/`sanitize` use the unicode `Character.isLetter`/`isLetterOrDigit`/`isDigit`, so Java keeps non-ASCII letters/digits as-is (`café`, leading `٠` → `_٠`, …) — which actually violates Avro's ASCII-only name grammar `[A-Za-z_][A-Za-z0-9_]*`. matching Java's `isDigit` here would re-introduce the exact invalid-Avro bug this PR fixes. so we stay ASCII-strict: anything outside ASCII alnum/`_` is escaped `_x<HEX>`, output is always spec-valid, original preserved in `iceberg-field-name`.

correction to the framing — it's **not** "both restore from the map". I checked the Java source: `iceberg-field-name` is **write-only** in Java (written by `TypeToSchema` + `BuildAvroProjection`; no read site). Java's `SchemaToType` just uses `field.name()`, and the authoritative name is recovered by field-id projection against the table schema. so correctness doesn't depend on the escape representation either way: within iceberg-rust the round-trip is lossless because we write *and* read the prop; cross-engine, names resolve by field-id. our read path honoring the prop is actually a touch more faithful than Java's `SchemaToType` on the standalone conversion path.

net: not changing the digit logic. I did fix the doc, which wrongly claimed it "matches Java semantics" — it follows Java's escape scheme but is intentionally ASCII-stricter.

**Comment by SreeramGarlapati** [comment]
@xanderbailey rebased onto latest main, ci green. i think both your points are covered above:
- `isDigit`: staying ascii-strict on purpose — matching Java's unicode `isDigit` would re-introduce the exact invalid-avro-name case this pr fixes, and cross-engine correctness is by field-id anyway, so representational parity isn't needed.
- avro-rs `validate_record_field_name`: kept the local 6-line check since it's zero-coupling and also covers the `iceberg-field-name` sanitize half that avro-rs doesn't. happy to file an upstream issue to expose their validator as a good-citizen follow-up if u want it.

anything else blocking here?
